Improve error message for invalid generic type arguments.

rfairfax opened the pull request:
Fix for issue #7177.  There are actually three separate issues that have come up on that thread that are in different areas of the code so I chose to focus on the initial report and let the other issues be addressed via other issues and commits.

Currently if we get an invalid generic type like Action<0> we end up generic 6-7 errors as the parser continues due to the fact that we start parsing the 0 as a brand new statement.  We now handle the case where the invalid type is followed by a comma or greater than and identify that we can continue parsing the type list to produce a much cleaner error state, now just the correct "Type expected" in many cases.

New unit tests included for these cases and the entire suite passes with the changes.

gafter commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5577,6 +5577,33 @@ private TypeSyntax ParseTypeArgument()
+                // for the majority of cases.
+                if (result.IsMissing &&
+                    (this.CurrentToken.Kind != SyntaxKind.CommaToken && this.CurrentToken.Kind != SyntaxKind.GreaterThanToken) && 
+                    (this.PeekToken(1).Kind == SyntaxKind.CommaToken || this.PeekToken(1).Kind == SyntaxKind.GreaterThanToken))
```
A type argument can also be followed by a `>>` token (e.g. in `List<List<.>>`), and a couple of other tokens as well. Should those be considered here?

gafter replied:
Oh, right, those are treated as separate tokens at this level.

rfairfax replied:
That's my understanding based on a review of the code (I'll get two tokens so it'll do the right thing as it's recursively parsing the type arguments list here) - but it's an excellent test case that is missing.  I'll add some tests for this.

AlekseyTs replied:
Is it safe to PeekTokens if we are at the end of the token stream?

AlekseyTs replied:
Also, consider using inline assignment to cache this.PeekToken(1).Kind in a local in the first part of the condition and reusing it in the second part.

rfairfax replied:
I reviewed the code for PeekToken and verified under a debugger - if you call PeekToken(n) with a value of n that goes beyond the end of the stream you get back an EOF token, so the method is always safe to call.

gafter commented:
The fix looks good, but I think the test coverage is a bit overkill. For example, it is not important that we maintain precisely the same parse trees for these kinds of error conditions in the future; the main point is to reduce the cascaded errors. That is best tested by checking the set of errors produced for a compilation unit. I think one or two tests each containing a compilation unit that contains a bunch of these kinds of errors, and checking for the resulting diagnostics produced by the parser, would be a good balance of test coverage.

rfairfax commented:
I 100% agree that the structure of the syntax tree is irrelevant here and that a change in the future isn't necessarily a defect.

Since this is a foreign code base to me I tried to find similar test cases and match them so that the testing strategies aligned given that I didn't have context on why certain tests were added originally and I tend to err on the side of over producing tests given that I don't have the context on if the full suite is run every time or sub-suites are used and the duplication in coverage then has some value.

Based on the feedback I'd propose the following test changes:

Keep things in NameParsingTests.cs - these are effectively the unit level tests for this logic at a very scoped parsing level making it easier to debug test failures.
Remove tests in StatementParsingTests.cs, ParsingErrorRecoveryTests.cs - these are redundant given the unit level coverage
Remove tests in SeparatedSyntaxListParsingTests - this test will largely just cause noise in the future.
Add a 1-2 larger compliation unit tests that covers a wide range of cases in one compilation and checks the diagnostics returned.  This forms the more E2E coverage for the change.  Any suggestion on what test class this should live in?

rfairfax commented:
I've adjusted the unit tests based on feedback - removed redundant tests, cut the syntax tree tests, and added a compilation unit test to the diagnostics suite that tries a number of variations on the repro input and checks the diagnostic output.

leppie commented:
@rfairfax: Nice fix, but from a compiler point of view, it is better just to test the result (aka `VerfiyDiagnostics` or verification when dealing with Emit). Any future improvements to syntax/semantics could cause pain for a contributor. Not saying your tests are bad, but as @gafter said, overkill.

leppie commented:
retest prtest/win/dbg/unit64  please

gafter commented:
@dotnet/roslyn-compiler Other reviewers care to chime in?

gafter commented:
LGTM.

@dotnet/roslyn-compiler Can we please have another review?

AlekseyTs commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -5577,6 +5577,33 @@ private TypeSyntax ParseTypeArgument()
+                // interpret 0 as the start of a new expression when we can tell it's most likely
+                // meant to be part of the type list.  
+                //
+                // To solve this we check if the current token is not comma or grater than and 
```
Typo "grater"?

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Diagnostics/DiagnosticTest.cs:
```diff
@@ -2061,6 +2061,66 @@ public static void Main()
+}";
+
+            var comp = DiagnosticsUtils.VerifyErrorsAndGetCompilationWithMscorlib(text,
+                // Case a1
```
Please, use baseline style as in PragmaWarning_IdentifierErrorCodesMustMatchExactly1 test. I.e. the error number, message text and corresponding source code is included as a comment. I suggest switching to using VerifyDiagnostics helper, this is the preferred helper to verify expected diagnostics, only two tests in this file are still using VerifyErrorsAndGetCompilationWithMscorlib helper. Also, if you use VerifyDiagnostics helper and run the test without supplying any arguments to it, the test failure output will contain baseline that you can simply copy  and paste as the argument list for VerifyDiagnostics.

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs:
```diff
@@ -113,7 +113,7 @@ public void TestDotGenericName()
             Assert.False(es.SemicolonToken.IsMissing);
         }
-
+        
```
Please revert this change.

AlekseyTs commented:
I also added a couple of comments for this line in this change https://github.com/dotnet/roslyn/pull/7624/files#diff-ac2b4121f27df06f497115cc68277654R5601

rfairfax commented:
The latest commit addresses all the feedback from @AlekseyTs.  As part of making test updates I made another fix related to the diagnostics and how I'm eating the bad token.  We now properly mark the point of the error at the start of the bad token and not the end of it.  For example, with the following bad line:

object a2 = typeof(Action<static>);

We were previously reporting the error / squiggling text after the 'c' in static.  We now properly show the error starts at the 's' in static and squiggle the whole word making it even more obvious to the user what the error is.

VSadov commented:
LGTM

AlekseyTs commented on src/Compilers/CSharp/Test/Syntax/Parsing/NameParsingTests.cs:
```diff
@@ -571,6 +571,90 @@ public void TestAttributeInNameBad()
+        [Fact]
+        public void TestConstantInGenericNameBad()
+        {
+            var text = "foo<0>";
```
It would be good to have tests for "foo<" and "foo<0" as well.
